### PR TITLE
fix(repl) recover from invalid unicode

### PR DESCRIPTION
**Pull request by benjamingr**
Fixes https://github.com/denoland/deno/issues/7953

This is not a very good fix (yet). The crux is that serde_json behaves differently from V8's JSON parser.

A better/more generic fix would be to "upstream" a fix in serde_json to allow recovery from this. In the meantime, I returned a result with the string "Unterminated string literal".

Mostly - doing this to practice rust and deno :]

### CODE REVIEW

**Comment by bartlomieju** [comment]
@caspervonb please take a look

**Comment by caspervonb** [comment]
Please add a test case with the bad inputs solson came up with.

**Comment by benjamingr** [comment]
@caspervonb I am happy to add a test - but given a test do the changes themselves and the approach (fixing this here until a fix is upstreamed to the json parsing code itself) look fine to you?

**Comment by caspervonb** [comment]
SGTM

**Comment by bnoordhuis** [comment]
Benjamin, just checking: is the issue here invalid Unicode or malformed JSON? Can you share an example message?

**Comment by benjamingr** [comment]
> Benjamin, just checking: is the issue here invalid Unicode or malformed JSON? Can you share an example message?

The issue here is that when you pass an unterminated unicode code unit to `serde_json` it will error but chrome's JSON parser happily processes it.

So when you enter something like `'\u{1f3b5}'[0]` in the Deno repl it will encode it in a websocket message and when Deno tries to deserialize it (in inspector.rs) the `.unwrap` call panics because serde_json does not deserialize it.

I am not sure if chrome or serde_json is right here in terms of the spec - this fix is just a patch that says "if you would have panicked - show an 'Unterminated string literal' error instead".

File: cli/inspector.rs [hunk 1]
```diff
@@ -856,8 +856,30 @@ impl v8::inspector::ChannelImpl for InspectorSession {
...
     let raw_message = message.unwrap().string().to_string();
-    let message = serde_json::from_str(&raw_message).unwrap();
+    let message: serde_json::Value = match serde_json::from_str(&raw_message) {
+      Ok(v) => v,
+      Err(error) => match error.classify() {
+        serde_json::error::Category::Syntax => json!({
+          "id": call_id,
+          "result": {
+            "result": {
+              "type": "error"
```
Comment (bartlomieju) [inline cli/inspector.rs:868]: missing comma here?

Comment (benjamingr) [reply cli/inspector.rs:868]: No this is JSON - can't put a comma here :]

Comment (benjamingr) [reply cli/inspector.rs:868]: I have no idea how this works the comma isn't optional this is JSON 😅 

good catch

**Comment by bnoordhuis** [comment]
> when you enter something like '\u{1f3b5}'[0] in the Deno repl it will encode it in a websocket message and when Deno tries to deserialize it (in inspector.rs) the .unwrap call panics because serde_json does not deserialize it.

Thanks, I'm able to reproduce using your example.

The inspector sends `{"id":3,"result":{"result":{"type":"string","value":"\ud83c"}}}` and `serde_json::from_str()` chokes on the unpaired surrogate (`U+D800 < U+D83C < U+DFFF`).

From `serde_json-1.0.59/src/de.rs`:
```rust
    /// The behavior of serde_json is specified to fail on non-UTF-8 strings
    /// when deserializing into Rust UTF-8 string types such as String, and
    /// succeed with non-UTF-8 bytes when deserializing using this method.
```
(Where "this method" is a helper to `serde_json::from_slice()`.)

Ergo, operating on `&[u8]` instead of `&str` should fix it but that means converting from `&[u16]` to `&[u8]` first and that's its own can of worms.

I've opened https://bugs.chromium.org/p/v8/issues/detail?id=11193 to discuss if it's better to handle this in V8. Deno is probably not the only embedder that has to grapple with this so I think it makes sense to fix this at the root.

**Comment by benjamingr** [comment]
@bnoordhuis yes though to be fair this _should_ probably be fixed in serde_json - Chrome's JSON parser happily deserializes the invalid JSON gracefully.

I'm currently bedridden recovering from a surprising sickness last weekend - I'll pick up this work in a few days when I feel better (hopefully) - if this is urgent for whatever reason feel free to pick this up.

**Comment by bnoordhuis** [comment]
🤷 - I don't think serde_json is wrong to reject it, else it would have to either emit invalid UTF-8 or transmogrify it into something like WTF-8.

Anyway, no rush. Get well soon! I can snail mail you some powdered chicken soup if it helps. :-)

**Comment by bartlomieju** [comment]
Fixed in #8759

**Pull request closed without merging**